File: source/pressure-gauge/dotnet/pressure_gauge_stepper_motor_p1.cs
// In this sample:
// Pressure (kPa) is read from a pressure sensor, and displayed on a gauge using stepper motor P1.

using GHIElectronics.DUELink;

var availablePort = DUELinkController.GetConnectionPort();
var duelink = new DUELinkController(availablePort);

var deviceAddress = 0;
var presssureValue = 0;
var currentGaugeValue = 0;

void SelectDevice(int address) {
    if (deviceAddress != address) {
        // Saving the bus, only update select when different current address
        deviceAddress = address;
        duelink.Engine.ExecuteCommand($"sel({deviceAddress})");
    }
}
void SetGauge(int value) {
    if (value == currentGaugeValue || value <0)
        return;

    SelectDevice(1);

    var target_step = StepFromValue(value);
    var current_step = StepFromValue(currentGaugeValue);

    var diff = target_step - current_step;

    var direction = diff >= 0 ? 1 : 0;

    duelink.Engine.ExecuteCommand($"step_m1({direction},{Math.Abs(diff)})");

    currentGaugeValue = value;
}

int ReadPressure() {
    // return kPa
    SelectDevice(2);
    var ret = (int)duelink.Engine.ExecuteCommand("kPa()");

    return ret;

}

int StepFromValue(int value) {
    // default resolution is 400 steps
    // reach to the value 100 on the gauge take 275 steps.
    return (int)(value * 2.75);
}


while (true) {
        
    var presssure_temp = ReadPressure();
    if (presssure_temp!= presssureValue) {
        presssureValue = presssure_temp;
        SetGauge(presssureValue);

        //Debug only
        Console.WriteLine($"Kpa reading: {presssureValue}");
    }

    Thread.Sleep(50);
}

File: source/pressure-gauge/py/pressure_gauge_stepper_motor_p1.py
#In this sample:
# Pressure (kPa) is read from a pressure sensor, and displayed on a gauge using stepper motor P1.

import time
from DUELink.DUELinkController import DUELinkController

#Enable this if connect by USB port
availablePort = DUELinkController.GetConnectionPort()
duelink = DUELinkController(availablePort) #Bluetooth COM port, change to user comport

# Linux
# duelink = DUELinkController("/dev/ttyAMA0")

#variable
deviceAddress = 0
presssureValue = 0
currentGaugeValue = 0

# methods
def SelectDevice(address):
    global deviceAddress

    if deviceAddress != address:
        # Save bus traffic; only update select when the address is different
        deviceAddress = address
        duelink.Engine.ExecuteCommand(f"sel({deviceAddress})")


def SetGauge(value):
    global currentGaugeValue

    if value == currentGaugeValue or value < 0:
        return

    SelectDevice(1)

    target_step = StepFromValue(value)
    current_step = StepFromValue(currentGaugeValue)

    diff = target_step - current_step

    direction = 1 if diff >= 0 else 0

    duelink.Engine.ExecuteCommand(f"step_m1({direction},{abs(diff)})")

    currentGaugeValue = value


def ReadPressure():
    # Return pressure in kPa
    SelectDevice(2)
    ret = int(duelink.Engine.ExecuteCommand("kPa()"))

    return ret


def StepFromValue(value):
    # Default resolution is 400 steps
    # Reaching the value 100 on the gauge requires 275 steps
    return int(value * 2.75)


while True:

    presssure_temp = ReadPressure()

    if presssure_temp != presssureValue:
        presssureValue = presssure_temp
        SetGauge(presssureValue)

        # Debug only
        print(f"Kpa reading: {presssureValue}")

    time.sleep(0.05)
